src/data/proxy_dmatrix.h

```diff
@@ -0,0 +1,112 @@
+/*!
+ * Copyright 2020 XGBoost contributors
+ */
+#ifndef XGBOOST_DATA_PROXY_DMATRIX_H_
+#define XGBOOST_DATA_PROXY_DMATRIX_H_
+
+#include <dmlc/any.h>
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "xgboost/data.h"
+#include "xgboost/generic_parameters.h"
+#include "xgboost/c_api.h"
+#include "adapter.h"
+
+namespace xgboost {
+namespace data {
+/*
+ * \brief A proxy to external iterator.
+ */
+template <typename ResetFn, typename NextFn>
+class DataIterProxy {
+  DataIterHandle iter_;
+  ResetFn* reset_;
+  NextFn* next_;
+
+ public:
+  DataIterProxy(DataIterHandle iter, ResetFn* reset, NextFn* next) :
+      iter_{iter},
+      reset_{reset}, next_{next} {}
+
+  bool Next() {
+    return next_(iter_);
+  }
+  void Reset() {
+    reset_(iter_);
+  }
+};
+
+/*
+ * \brief A proxy of DMatrix used by external iterator.
+ */
+class DMatrixProxy : public DMatrix {
+  MetaInfo info_;
+  dmlc::any batch_;
+  int32_t device_ { xgboost::GenericParameter::kCpuId };
+
+ public:
+  void SetInfo(char const *key, void const *info, DataType type,
+               size_t len) override {
+    this->Info().SetInfo(key, info, type, len);
+  }
+  void SetInfo(const char* key, std::string const& interface_str) override {
+    this->Info().SetInfo(key, interface_str);
+  }
+
+  int DeviceIdx() const { return device_; }
+
+#if defined(XGBOOST_USE_CUDA)
+  void FromCudaColumnar(std::string interface_str);
+  void FromCudaArray(std::string interface_str);
+#endif  // defined(XGBOOST_USE_CUDA)
+
+  void SetData(char const* c_interface) {
+    common::AssertGPUSupport();
```

Comment:
Is `DMatrixProxy::SetData()` exclusive to GPUs? The name `DMatrixProxy` doesn't suggest that GPU is needed.

Reply:
I'm trying to keep everything general, so that later it can be reused by others like CPU and sycl.